org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractHoverTest.java

```diff
@@ -0,0 +1,129 @@
+ */
+@SuppressWarnings("restriction")
+public abstract class AbstractHoverTest extends AbstractEditorTest {
+
+	@Inject
+	protected IEObjectHover eObjectHover;
+
+	@Inject
+	protected SyncUtil syncUtil;
+
+	@Inject
+	protected FileExtensionProvider fileExtensionProvider;
+
+	/**
+	 * Test that the expected text appears in a popup window while hovering over a specific text in a given DSL text.
+	 * 
+	 * @param charSequence
+	 *            the initial DSL text.
+	 * @param hoverText
+	 *            the text you are hovering over.
+	 * @param hoverContent
+	 *            the text that is expected to be part of the hover popup.
+	 */
+	public void hasHoverOver(CharSequence charSequence, String hoverText, String hoverContent) {
+		Region region = new Region(charSequence.toString().indexOf(hoverText), hoverText.length());
+		hasHoverOver(charSequence, region, hoverContent);
+	}
+
+	/**
+	 * Test that the expected text appears in a popup window while hovering over a specific region in a given DSL text.
+	 * 
+	 * @param it
+	 *            the initial DSL text.
+	 * @param hoverRegion
+	 *            the region you are hovering over.
+	 * @param hoverContent
+	 *            the text that is expected to be part of the hover popup.
+	 */
+	public void hasHoverOver(CharSequence charSequence, IRegion hoverRegion, String hoverContent) {
+		hoverPopupHasContent(hoveringOver(dslFile(charSequence), hoverRegion), hoverContent);
+	}
+
+	protected IFile dslFile(CharSequence content) {
+		try {
+			IFile file = IResourcesSetupUtil.createFile(getProjectName(), getFileName(), getFileExtension(), content.toString());
+			IProject project = file.getProject();
+			if (!project.hasNature(XtextProjectHelper.NATURE_ID)) {
+				IResourcesSetupUtil.addNature(project, XtextProjectHelper.NATURE_ID);
+			}
+			return file;
+		} catch (InvocationTargetException | CoreException | InterruptedException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected String getProjectName() {
+		return "HoverTestProject";
+	}
+
+	protected String getFileName() {
+		return "hover";
+	}
+
+	protected String getFileExtension() {
+		return fileExtensionProvider.getPrimaryFileExtension();
+	}
+
+	protected XtextEditor openInEditor(IFile dslFile) {
+		try {
+			syncUtil.waitForBuild(new NullProgressMonitor());
+			return openEditor(dslFile);
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	protected BrowserInformationControlInput hoveringOver(IFile dslFile, IRegion hoverRegion) {
+		XtextEditor editor = openInEditor(dslFile);
+		Object hoverInfo = ((ITextHoverExtension2) eObjectHover).getHoverInfo2(editor.getInternalSourceViewer(), hoverRegion);
+		return (BrowserInformationControlInput) hoverInfo;
+	}
+
+	protected void hoverPopupHasContent(BrowserInformationControlInput info, String expectedHoverContent) {
+		assertNotNull("No hover found!", info);
+
+		String actualHoverContent = info.getHtml();
+		StringConcatenation builder = new StringConcatenation();
```

Comment:
I am not sure how I could replace the call to `#newLineIfNotEmpty()`, which is why I have kept this code as it is.